File: src/my_project/interviews/top_150_questions_round_22/ex_140_longest_palindromic_substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        start, max_len = 0, 1

        def expand(left: int, right: int) -> None:
            nonlocal start, max_len
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            length = right - left - 1
            if length > max_len:
                max_len = length
                start = left + 1

        for i in range(len(s)):
            expand(i, i)      # odd-length
            expand(i, i + 1)  # even-length

        return s[start:start + max_len]

File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_140_longest_palindromic_substring.ts
function longestPalindrome(s: string): string {
    let start = 0, maxLen = 1;

    const expand = (left: number, right: number): void => {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            left--;
            right++;
        }
        const length = right - left - 1;
        if (length > maxLen) {
            maxLen = length;
            start = left + 1;
        }
    };

    for (let i = 0; i < s.length; i++) {
        expand(i, i);       // odd-length
        expand(i, i + 1);   // even-length
    }

    return s.substring(start, start + maxLen);
}
